// @moonshot-ai/kimi-agent-sdk/dist/errors.mjs
// errors.ts
var TransportErrorCodes = {
  SPAWN_FAILED: "SPAWN_FAILED",
  STDIN_NOT_WRITABLE: "STDIN_NOT_WRITABLE",
  PROCESS_CRASHED: "PROCESS_CRASHED",
  CLI_NOT_FOUND: "CLI_NOT_FOUND",
  ALREADY_STARTED: "ALREADY_STARTED",
  HANDSHAKE_TIMEOUT: "HANDSHAKE_TIMEOUT"
};
var ProtocolErrorCodes = {
  INVALID_JSON: "INVALID_JSON",
  INVALID_REQUEST: "INVALID_REQUEST",
  INVALID_PARAMS: "INVALID_PARAMS",
  INTERNAL_ERROR: "INTERNAL_ERROR",
  SCHEMA_MISMATCH: "SCHEMA_MISMATCH",
  UNKNOWN_EVENT_TYPE: "UNKNOWN_EVENT_TYPE",
  UNKNOWN_REQUEST_TYPE: "UNKNOWN_REQUEST_TYPE",
  REQUEST_TIMEOUT: "REQUEST_TIMEOUT",
  REQUEST_CANCELLED: "REQUEST_CANCELLED"
};
var SessionErrorCodes = {
  SESSION_CLOSED: "SESSION_CLOSED",
  SESSION_BUSY: "SESSION_BUSY",
  TURN_INTERRUPTED: "TURN_INTERRUPTED",
  APPROVAL_FAILED: "APPROVAL_FAILED"
};
var CliErrorCodes = {
  INVALID_STATE: "INVALID_STATE",
  LLM_NOT_SET: "LLM_NOT_SET",
  LLM_NOT_SUPPORTED: "LLM_NOT_SUPPORTED",
  CHAT_PROVIDER_ERROR: "CHAT_PROVIDER_ERROR",
  UNKNOWN: "UNKNOWN"
};
var AgentSdkError = class extends Error {
  constructor(message, cause, context) {
    super(message);
    this.cause = cause;
    this.context = context;
    this.name = this.constructor.name;
  }
};
var TransportError = class extends AgentSdkError {
  constructor(code, message, cause) {
    super(message, cause);
    this.code = code;
  }
  category = "transport";
};
var ProtocolError = class extends AgentSdkError {
  constructor(code, message, context) {
    super(message, void 0, context);
    this.code = code;
  }
  category = "protocol";
};
var SessionError = class extends AgentSdkError {
  constructor(code, message) {
    super(message);
    this.code = code;
  }
  category = "session";
};
var CliError = class _CliError extends AgentSdkError {
  constructor(code, message, numericCode, rawResponse) {
    super(message);
    this.code = code;
    this.numericCode = numericCode;
    this.rawResponse = rawResponse;
  }
  category = "cli";
  static fromRpcError(rpcCode, message, rawJson) {
    const protocolCodeMap = {
      [-32700]: ProtocolErrorCodes.INVALID_JSON,
      [-32600]: ProtocolErrorCodes.INVALID_REQUEST,
      [-32602]: ProtocolErrorCodes.INVALID_PARAMS,
      [-32603]: ProtocolErrorCodes.INTERNAL_ERROR
    };
    if (protocolCodeMap[rpcCode]) {
      return new ProtocolError(protocolCodeMap[rpcCode], message);
    }
    const cliCodeMap = {
      [-32e3]: CliErrorCodes.INVALID_STATE,
      [-32001]: CliErrorCodes.LLM_NOT_SET,
      [-32002]: CliErrorCodes.LLM_NOT_SUPPORTED,
      [-32003]: CliErrorCodes.CHAT_PROVIDER_ERROR
    };
    return new _CliError(cliCodeMap[rpcCode] ?? CliErrorCodes.UNKNOWN, message, rpcCode, rawJson);
  }
};
function isAgentSdkError(err) {
  return err instanceof AgentSdkError;
}
function getErrorCode(err) {
  if (isAgentSdkError(err)) {
    return err.code;
  }
  return "UNKNOWN";
}
function getErrorCategory(err) {
  if (isAgentSdkError(err)) {
    return err.category;
  }
  return "unknown";
}

export { AgentSdkError, CliError, CliErrorCodes, ProtocolError, ProtocolErrorCodes, SessionError, SessionErrorCodes, TransportError, TransportErrorCodes, getErrorCategory, getErrorCode, isAgentSdkError };
