// @moonshot-ai/kimi-agent-sdk/dist/index.mjs
import * as crypto from 'crypto';
import { spawn } from 'child_process';
import * as readline from 'readline';
import { createInterface } from 'readline';
import { z } from 'zod';
import createDebug from 'debug';
import * as fs from 'fs';
import * as fsp from 'fs/promises';
import * as path2 from 'path';
import * as os from 'os';
import * as toml from 'toml';
import { zodToJsonSchema } from 'zod-to-json-schema';

// session.ts
var ApprovalResponseSchema = z.enum(["approve", "approve_for_session", "reject"]);
var ContentPartSchema = z.discriminatedUnion("type", [
  z.object({
    // Text type
    type: z.literal("text"),
    // Text content
    text: z.string()
  }),
  z.object({
    // Think type, appears only in thinking mode
    type: z.literal("think"),
    // Think content
    think: z.string(),
    // Encrypted think content or signature
    encrypted: z.string().nullable().optional()
  }),
  z.object({
    // Image type
    type: z.literal("image_url"),
    image_url: z.object({
      // Image URL, usually a data URI (e.g., data:image/png;base64,...)
      url: z.string(),
      // Image ID, used to distinguish different images
      id: z.string().nullable().optional()
    })
  }),
  z.object({
    // Audio type
    type: z.literal("audio_url"),
    audio_url: z.object({
      // Audio URL, usually a data URI (e.g., data:audio/aac;base64,...)
      url: z.string(),
      // Audio ID, used to distinguish different audio
      id: z.string().nullable().optional()
    })
  }),
  z.object({
    // Video type
    type: z.literal("video_url"),
    video_url: z.object({
      // Video URL, usually a data URI (e.g., data:video/mp4;base64,...)
      url: z.string(),
      // Video ID, used to distinguish different videos
      id: z.string().nullable().optional()
    })
  })
]);
var TokenUsageSchema = z.object({
  // Number of input tokens (excluding cache)
  input_other: z.number(),
  // Number of output tokens
  output: z.number(),
  // Number of input tokens read from cache
  input_cache_read: z.number(),
  // Number of input tokens written to cache
  input_cache_creation: z.number()
});
z.object({
  type: z.literal("brief"),
  // Brief text content
  text: z.string()
});
z.object({
  type: z.literal("diff"),
  // File path
  path: z.string(),
  // Original content
  old_text: z.string(),
  // New content
  new_text: z.string()
});
z.object({
  type: z.literal("todo"),
  // Todo list
  items: z.array(
    z.object({
      // Todo item title
      title: z.string(),
      // Status: pending | in_progress | done
      status: z.enum(["pending", "in_progress", "done"])
    })
  )
});
z.object({
  type: z.literal("shell"),
  language: z.string(),
  command: z.string()
});
var RawDisplayBlockSchema = z.object({
  type: z.string(),
  text: z.string().optional(),
  path: z.string().optional(),
  old_text: z.string().optional(),
  new_text: z.string().optional(),
  items: z.array(z.object({ title: z.string(), status: z.enum(["pending", "in_progress", "done"]) })).optional(),
  language: z.string().optional(),
  command: z.string().optional()
}).passthrough();
var DisplayBlockSchema = RawDisplayBlockSchema.transform((raw) => {
  switch (raw.type) {
    case "brief":
      if (typeof raw.text === "string") {
        return { type: "brief", text: raw.text };
      }
      break;
    case "diff":
      if (typeof raw.path === "string" && typeof raw.old_text === "string" && typeof raw.new_text === "string") {
        return { type: "diff", path: raw.path, old_text: raw.old_text, new_text: raw.new_text };
      }
      break;
    case "todo":
      if (Array.isArray(raw.items)) {
        return { type: "todo", items: raw.items };
      }
      break;
    case "shell":
      if (typeof raw.language === "string" && typeof raw.command === "string") {
        return { type: "shell", language: raw.language, command: raw.command };
      }
      break;
  }
  const { type, ...rest } = raw;
  return { type, data: rest };
});
var ToolCallSchema = z.object({
  // Always "function"
  type: z.literal("function"),
  // Tool call ID, used to match with ToolResult
  id: z.string(),
  function: z.object({
    // Tool name, e.g. "Shell", "ReadFile", "WriteFile"
    name: z.string(),
    // JSON-encoded arguments string, may be incomplete during streaming
    arguments: z.string().nullable().optional()
  }),
  // Extra metadata
  extras: z.record(z.string(), z.unknown()).nullable().optional()
});
var ToolCallPartSchema = z.object({
  // Argument chunk, appended to the last ToolCall's arguments
  arguments_part: z.string().nullable().optional()
});
var ToolReturnValueSchema = z.object({
  is_error: z.boolean(),
  output: z.union([z.string(), z.array(ContentPartSchema)]),
  message: z.string(),
  display: z.array(DisplayBlockSchema),
  extras: z.record(z.string(), z.unknown()).nullable().optional()
});
var ToolResultSchema = z.object({
  // Corresponding tool call ID
  tool_call_id: z.string(),
  return_value: ToolReturnValueSchema
});
var ClientInfoSchema = z.object({
  name: z.string(),
  version: z.string().optional()
});
var ServerInfoSchema = z.object({
  name: z.string(),
  version: z.string()
});
var SlashCommandInfoSchema = z.object({
  name: z.string(),
  description: z.string(),
  aliases: z.array(z.string())
});
var ExternalToolDefinitionSchema = z.object({
  name: z.string(),
  description: z.string(),
  parameters: z.record(z.string(), z.unknown())
});
var ExternalToolsResultSchema = z.object({
  accepted: z.array(z.string()),
  rejected: z.array(z.object({ name: z.string(), reason: z.string() }))
});
var ClientCapabilitiesSchema = z.object({
  // Whether the client supports handling QuestionRequest messages
  supports_question: z.boolean().optional()
});
var ServerCapabilitiesSchema = z.object({
  // Whether the server supports sending QuestionRequest messages
  supports_question: z.boolean().optional()
});
z.object({
  protocol_version: z.string(),
  client: ClientInfoSchema.optional(),
  external_tools: z.array(ExternalToolDefinitionSchema).optional(),
  capabilities: ClientCapabilitiesSchema.optional()
});
var InitializeResultSchema = z.object({
  protocol_version: z.string(),
  server: ServerInfoSchema,
  slash_commands: z.array(SlashCommandInfoSchema),
  external_tools: ExternalToolsResultSchema.optional(),
  capabilities: ServerCapabilitiesSchema.optional()
});
var ToolCallRequestSchema = z.object({
  id: z.string(),
  name: z.string(),
  arguments: z.string().nullable().optional()
});
var QuestionOptionSchema = z.object({
  // Option label
  label: z.string(),
  // Option description
  description: z.string().optional()
});
var QuestionItemSchema = z.object({
  // Question text
  question: z.string(),
  // Short header label (max 12 chars)
  header: z.string().optional(),
  // Options (2-4)
  options: z.array(QuestionOptionSchema),
  // Whether multiple selection is allowed
  multi_select: z.boolean().optional()
});
var QuestionRequestSchema = z.object({
  // Request ID, used for response reference
  id: z.string(),
  // Associated tool call ID
  tool_call_id: z.string(),
  // Questions (1-4)
  questions: z.array(QuestionItemSchema)
});
z.object({
  // Corresponding request ID
  request_id: z.string(),
  // Answer mapping: question text -> selected option label(s)
  answers: z.record(z.string(), z.string())
});
var TurnBeginSchema = z.object({
  // User input, can be plain text or array of content parts
  user_input: z.union([z.string(), z.array(ContentPartSchema)])
});
var StepBeginSchema = z.object({
  // Step number, starting from 1
  n: z.number()
});
var EmptyPayloadSchema = z.object({});
var StatusUpdateSchema = z.object({
  // Context usage ratio, float between 0 and 1
  context_usage: z.number().nullable().optional(),
  // Token usage stats for the current step
  token_usage: TokenUsageSchema.nullable().optional(),
  // Message ID for the current step
  message_id: z.string().nullable().optional()
});
var ApprovalRequestPayloadSchema = z.object({
  // Request ID, referenced when responding
  id: z.string(),
  // Associated tool call ID
  tool_call_id: z.string(),
  // Sender (tool name), e.g. "Shell", "WriteFile"
  sender: z.string(),
  // Action description, e.g. "run shell command"
  action: z.string(),
  // Detailed description, e.g. "Run command `rm -rf /`"
  description: z.string(),
  // Display blocks shown to the user
  display: z.array(DisplayBlockSchema).optional()
});
var ApprovalResponseEventSchema = z.object({
  // Resolved approval request ID
  request_id: z.string(),
  // Approval result
  response: ApprovalResponseSchema
});
var EventSchemas = {
  TurnBegin: TurnBeginSchema,
  TurnEnd: EmptyPayloadSchema,
  StepBegin: StepBeginSchema,
  StepInterrupted: EmptyPayloadSchema,
  CompactionBegin: EmptyPayloadSchema,
  CompactionEnd: EmptyPayloadSchema,
  StatusUpdate: StatusUpdateSchema,
  ContentPart: ContentPartSchema,
  ToolCall: ToolCallSchema,
  ToolCallPart: ToolCallPartSchema,
  ToolResult: ToolResultSchema,
  ApprovalResponse: ApprovalResponseEventSchema
};
var RequestSchemas = {
  ApprovalRequest: ApprovalRequestPayloadSchema,
  ToolCallRequest: ToolCallRequestSchema,
  QuestionRequest: QuestionRequestSchema
};
function parseEventPayload(type, payload) {
  const schema = EventSchemas[type];
  if (!schema) {
    return { ok: false, error: `Unknown event type: ${type}` };
  }
  const result = schema.safeParse(payload);
  if (!result.success) {
    return { ok: false, error: `Invalid payload for ${type}: ${result.error.message}` };
  }
  return { ok: true, value: { type, payload: result.data } };
}
function parseWireEvent(raw) {
  const result = parseEventPayload(raw.type, raw.payload);
  if (result.ok) {
    return result.value;
  }
  return {
    type: "ParseError",
    payload: {
      code: "SUBAGENT_PARSE_FAILED",
      message: result.error,
      rawType: raw.type
    }
  };
}
var SubagentEventSchema = z.lazy(
  () => z.object({
    task_tool_call_id: z.string(),
    event: z.object({ type: z.string(), payload: z.unknown() }).transform(parseWireEvent)
  })
);
EventSchemas.SubagentEvent = SubagentEventSchema;
var RunResultSchema = z.object({
  /**
   * Completion status
   * - `finished`: Turn completed normally
   * - `cancelled`: Turn was cancelled via cancel()
   * - `max_steps_reached`: Reached maximum step limit
   */
  status: z.enum(["finished", "cancelled", "max_steps_reached"]),
  // When status is max_steps_reached, returns the number of steps executed
  steps: z.number().optional()
});
var ReplayResultSchema = z.object({
  /**
   * Replay completion status
   * - `finished`: Replay completed normally
   * - `cancelled`: Replay was cancelled via cancel()
   */
  status: z.enum(["finished", "cancelled"]),
  // Number of events replayed
  events: z.number(),
  // Number of requests replayed
  requests: z.number()
});
var RpcErrorSchema = z.object({
  // Error code
  code: z.number(),
  // Error message
  message: z.string(),
  // Extra data
  data: z.unknown().optional()
});
z.object({
  jsonrpc: z.string().optional(),
  id: z.string().optional(),
  method: z.string().optional(),
  params: z.unknown().optional(),
  result: z.unknown().optional(),
  error: RpcErrorSchema.optional()
});
function parseRequestPayload(type, payload) {
  const schema = RequestSchemas[type];
  if (!schema) {
    return { ok: false, error: `Unknown request type: ${type}` };
  }
  const result = schema.safeParse(payload);
  if (!result.success) {
    return { ok: false, error: `Invalid payload for ${type}: ${result.error.message}` };
  }
  return { ok: true, value: { type, payload: result.data } };
}
z.object({
  role: z.string().optional(),
  content: z.unknown().optional(),
  tool_calls: z.array(
    z.object({
      id: z.string().optional(),
      function: z.object({
        name: z.string().optional(),
        arguments: z.union([z.string(), z.record(z.string(), z.unknown())]).optional()
      }).optional()
    })
  ).optional(),
  tool_call_id: z.string().optional()
});

// errors.ts
var TransportErrorCodes = {
  SPAWN_FAILED: "SPAWN_FAILED",
  STDIN_NOT_WRITABLE: "STDIN_NOT_WRITABLE",
  PROCESS_CRASHED: "PROCESS_CRASHED",
  CLI_NOT_FOUND: "CLI_NOT_FOUND",
  ALREADY_STARTED: "ALREADY_STARTED",
  HANDSHAKE_TIMEOUT: "HANDSHAKE_TIMEOUT"
};
var ProtocolErrorCodes = {
  INVALID_JSON: "INVALID_JSON",
  INVALID_REQUEST: "INVALID_REQUEST",
  INVALID_PARAMS: "INVALID_PARAMS",
  INTERNAL_ERROR: "INTERNAL_ERROR",
  SCHEMA_MISMATCH: "SCHEMA_MISMATCH",
  UNKNOWN_EVENT_TYPE: "UNKNOWN_EVENT_TYPE",
  UNKNOWN_REQUEST_TYPE: "UNKNOWN_REQUEST_TYPE",
  REQUEST_TIMEOUT: "REQUEST_TIMEOUT",
  REQUEST_CANCELLED: "REQUEST_CANCELLED"
};
var SessionErrorCodes = {
  SESSION_CLOSED: "SESSION_CLOSED",
  SESSION_BUSY: "SESSION_BUSY",
  TURN_INTERRUPTED: "TURN_INTERRUPTED",
  APPROVAL_FAILED: "APPROVAL_FAILED"
};
var CliErrorCodes = {
  INVALID_STATE: "INVALID_STATE",
  LLM_NOT_SET: "LLM_NOT_SET",
  LLM_NOT_SUPPORTED: "LLM_NOT_SUPPORTED",
  CHAT_PROVIDER_ERROR: "CHAT_PROVIDER_ERROR",
  UNKNOWN: "UNKNOWN"
};
var AgentSdkError = class extends Error {
  constructor(message, cause, context) {
    super(message);
    this.cause = cause;
    this.context = context;
    this.name = this.constructor.name;
  }
};
var TransportError = class extends AgentSdkError {
  constructor(code, message, cause) {
    super(message, cause);
    this.code = code;
  }
  category = "transport";
};
var ProtocolError = class extends AgentSdkError {
  constructor(code, message, context) {
    super(message, void 0, context);
    this.code = code;
  }
  category = "protocol";
};
var SessionError = class extends AgentSdkError {
  constructor(code, message) {
    super(message);
    this.code = code;
  }
  category = "session";
};
var CliError = class _CliError extends AgentSdkError {
  constructor(code, message, numericCode, rawResponse) {
    super(message);
    this.code = code;
    this.numericCode = numericCode;
    this.rawResponse = rawResponse;
  }
  category = "cli";
  static fromRpcError(rpcCode, message, rawJson) {
    const protocolCodeMap = {
      [-32700]: ProtocolErrorCodes.INVALID_JSON,
      [-32600]: ProtocolErrorCodes.INVALID_REQUEST,
      [-32602]: ProtocolErrorCodes.INVALID_PARAMS,
      [-32603]: ProtocolErrorCodes.INTERNAL_ERROR
    };
    if (protocolCodeMap[rpcCode]) {
      return new ProtocolError(protocolCodeMap[rpcCode], message);
    }
    const cliCodeMap = {
      [-32e3]: CliErrorCodes.INVALID_STATE,
      [-32001]: CliErrorCodes.LLM_NOT_SET,
      [-32002]: CliErrorCodes.LLM_NOT_SUPPORTED,
      [-32003]: CliErrorCodes.CHAT_PROVIDER_ERROR
    };
    return new _CliError(cliCodeMap[rpcCode] ?? CliErrorCodes.UNKNOWN, message, rpcCode, rawJson);
  }
};
function isAgentSdkError(err) {
  return err instanceof AgentSdkError;
}
function getErrorCode(err) {
  if (isAgentSdkError(err)) {
    return err.code;
  }
  return "UNKNOWN";
}
function getErrorCategory(err) {
  if (isAgentSdkError(err)) {
    return err.category;
  }
  return "unknown";
}
var NAMESPACE = "kimi-sdk";
var log = {
  protocol: createDebug(`${NAMESPACE}:protocol`),
  session: createDebug(`${NAMESPACE}:session`),
  storage: createDebug(`${NAMESPACE}:storage`),
  config: createDebug(`${NAMESPACE}:config`),
  cli: createDebug(`${NAMESPACE}:cli`),
  history: createDebug(`${NAMESPACE}:history`)
};
function enableLogs(namespaces = `${NAMESPACE}:*`) {
  createDebug.enable(namespaces);
}
function disableLogs() {
  createDebug.disable();
}
function setLogSink(sink) {
  createDebug.log = sink;
}

// protocol.ts
var PROTOCOL_VERSION = "1.4";
var SDK_NAME = "kimi-agent-sdk";
var SDK_VERSION = "0.1.5" ;
function createEventChannel() {
  const queue = [];
  const resolvers = [];
  let finished = false;
  return {
    iterable: {
      [Symbol.asyncIterator]: () => ({
        next: () => {
          const queued = queue.shift();
          if (queued !== void 0) {
            return Promise.resolve({ done: false, value: queued });
          }
          if (finished) {
            return Promise.resolve({ done: true, value: void 0 });
          }
          return new Promise((resolve) => resolvers.push(resolve));
        }
      })
    },
    push: (value) => {
      if (finished) {
        return;
      }
      const resolver = resolvers.shift();
      if (resolver) {
        resolver({ done: false, value });
      } else {
        queue.push(value);
      }
    },
    finish: () => {
      if (finished) {
        return;
      }
      finished = true;
      for (const resolver of resolvers) {
        resolver({ done: true, value: void 0 });
      }
      resolvers.length = 0;
    }
  };
}
var ProtocolClient = class {
  process = null;
  readline = null;
  requestId = 0;
  stderrBuffer = "";
  pendingRequests = /* @__PURE__ */ new Map();
  pushEvent = null;
  finishEvents = null;
  externalToolHandlers = /* @__PURE__ */ new Map();
  get isRunning() {
    return this.process !== null && this.process.exitCode === null;
  }
  async start(options) {
    if (this.process) {
      throw new TransportError("ALREADY_STARTED", "Client already started");
    }
    const args = this.buildArgs(options);
    const executable = options.executablePath ?? "kimi";
    log.protocol("Spawning CLI: %s %o", executable, args);
    try {
      const { DEBUG, ...cleanEnv } = process.env;
      this.process = spawn(executable, args, {
        cwd: options.workDir,
        env: { ...cleanEnv, ...options.environmentVariables },
        stdio: ["pipe", "pipe", "pipe"]
      });
    } catch (err) {
      throw new TransportError("SPAWN_FAILED", `Failed to spawn CLI: ${err}`, err);
    }
    if (!this.process.stdout || !this.process.stdin) {
      this.process.kill();
      this.process = null;
      throw new TransportError("SPAWN_FAILED", "Process missing stdio");
    }
    this.readline = createInterface({ input: this.process.stdout });
    this.readline.on("line", (line) => this.handleLine(line));
    this.process.stderr?.on("data", (data) => {
      const chunk = data.toString();
      this.stderrBuffer += chunk;
      log.protocol("stderr: %s", chunk.trim());
    });
    this.process.on("error", (err) => this.handleProcessError(err));
    this.process.on("exit", (code) => this.handleProcessExit(code));
    if (options.externalTools) {
      for (const tool of options.externalTools) {
        this.externalToolHandlers.set(tool.name, tool.handler);
      }
    }
    const initResult = await this.sendInitialize(options.externalTools, options.clientInfo);
    return initResult;
  }
  async stop() {
    if (!this.process) {
      return;
    }
    if (this.process.exitCode !== null || this.process.killed) {
      this.cleanup();
      return;
    }
    log.protocol("Stopping process...");
    this.process.kill("SIGTERM");
    await new Promise((resolve) => {
      const timeout = setTimeout(() => {
        this.process?.kill("SIGKILL");
        resolve();
      }, 3e3);
      this.process.once("exit", () => {
        clearTimeout(timeout);
        resolve();
      });
    });
    this.cleanup();
  }
  sendPrompt(content) {
    const { iterable, push, finish } = createEventChannel();
    this.pushEvent = push;
    this.finishEvents = () => {
      finish();
      this.pushEvent = null;
      this.finishEvents = null;
    };
    const result = this.sendRequest("prompt", { user_input: content }).then((res) => {
      this.finishEvents?.();
      const r = res;
      return { status: r.status, steps: r.steps };
    }).catch((err) => {
      this.finishEvents?.();
      throw err;
    });
    return { events: iterable, result };
  }
  sendCancel() {
    return this.sendRequest("cancel").then(() => {
    });
  }
  sendReplay() {
    const { iterable, push, finish } = createEventChannel();
    this.pushEvent = push;
    this.finishEvents = () => {
      finish();
      this.pushEvent = null;
      this.finishEvents = null;
    };
    const result = this.sendRequest("replay").then((res) => {
      this.finishEvents?.();
      const parsed = ReplayResultSchema.safeParse(res);
      if (!parsed.success) {
        throw new ProtocolError("SCHEMA_MISMATCH", `Invalid replay response: ${parsed.error.message}`);
      }
      return parsed.data;
    }).catch((err) => {
      this.finishEvents?.();
      throw err;
    });
    return { events: iterable, result };
  }
  sendApproval(requestId, response) {
    this.writeLine({
      jsonrpc: "2.0",
      id: requestId,
      result: { request_id: requestId, response }
    });
    return Promise.resolve();
  }
  sendQuestionResponse(requestId, questionRequestId, answers) {
    this.writeLine({
      jsonrpc: "2.0",
      id: requestId,
      result: { request_id: questionRequestId, answers }
    });
    return Promise.resolve();
  }
  async sendInitialize(externalTools, clientInfo) {
    let clientName = `${SDK_NAME}/${SDK_VERSION}`;
    if (clientInfo?.name && clientInfo?.version) {
      clientName += ` ${clientInfo.name}/${clientInfo.version}`;
    }
    const params = {
      protocol_version: PROTOCOL_VERSION,
      client: {
        name: clientName,
        version: SDK_VERSION
      },
      capabilities: {
        supports_question: true
      }
    };
    if (externalTools && externalTools.length > 0) {
      params.external_tools = externalTools.map((tool) => ({
        name: tool.name,
        description: tool.description,
        parameters: tool.parameters
      }));
    }
    log.protocol("Sending initialize request: %O", params);
    const result = await this.sendRequest("initialize", params);
    const parsed = InitializeResultSchema.safeParse(result);
    log.protocol("Received initialize response: %O", result);
    if (!parsed.success) {
      throw new TransportError("SPAWN_FAILED", `Invalid initialize response: ${parsed.error.message}`);
    }
    log.protocol("Initialized: protocol=%s, server=%s/%s", parsed.data.protocol_version, parsed.data.server.name, parsed.data.server.version);
    return parsed.data;
  }
  // Private: Args Building
  buildArgs(options) {
    const args = [];
    if (options.sessionId) {
      args.push("--session", options.sessionId);
    }
    args.push("--work-dir", options.workDir, "--wire");
    if (options.model) {
      args.push("--model", options.model);
    }
    if (options.thinking) {
      args.push("--thinking");
    } else {
      args.push("--no-thinking");
    }
    if (options.yoloMode) {
      args.push("--yolo");
    }
    if (options.agentFile) {
      args.push("--agent-file", options.agentFile);
    }
    if (options.skillsDir) {
      args.push("--skills-dir", options.skillsDir);
    }
    return args;
  }
  // Private: RPC Communication
  sendRequest(method, params) {
    const id = `${++this.requestId}_${Date.now()}`;
    return new Promise((resolve, reject) => {
      this.pendingRequests.set(id, { resolve, reject });
      try {
        this.writeLine({ jsonrpc: "2.0", id, method, ...params && { params } });
      } catch (err) {
        this.pendingRequests.delete(id);
        reject(err);
      }
    });
  }
  writeLine(data) {
    log.protocol(">>> %O", data);
    if (!this.process?.stdin?.writable) {
      throw new TransportError("STDIN_NOT_WRITABLE", "Cannot write to CLI stdin");
    }
    this.process.stdin.write(JSON.stringify(data) + "\n");
  }
  // Private: Line Handling
  handleLine(line) {
    log.protocol("<<< %s", line.length > 500 ? line.slice(0, 500) + "..." : line);
    let parsed;
    try {
      parsed = JSON.parse(line);
    } catch {
      this.emitParseError("INVALID_JSON", "Failed to parse JSON", line);
      return;
    }
    const msg = parsed;
    if (msg.id && this.pendingRequests.has(msg.id)) {
      const pending = this.pendingRequests.get(msg.id);
      this.pendingRequests.delete(msg.id);
      if (msg.error) {
        pending.reject(CliError.fromRpcError(msg.error.code, msg.error.message, JSON.stringify(msg)));
      } else {
        pending.resolve(msg.result);
      }
      return;
    }
    if (msg.method) {
      if (msg.method === "request" && msg.id) {
        this.handleServerRequest(msg.id, msg.params);
      } else {
        this.handleNotification(msg.method, msg.params);
      }
    }
  }
  handleNotification(method, params) {
    if (method === "event") {
      const p = params;
      if (!p?.type) {
        this.emitParseError("SCHEMA_MISMATCH", "Event missing type");
        return;
      }
      const result = parseEventPayload(p.type, p.payload);
      if (result.ok) {
        this.pushEvent?.(result.value);
      } else {
        this.emitParseError("UNKNOWN_EVENT_TYPE", result.error);
      }
    }
  }
  handleServerRequest(requestId, params) {
    const p = params;
    if (!p?.type) {
      this.emitParseError("SCHEMA_MISMATCH", "Request missing type");
      return;
    }
    if (p.type === "ToolCallRequest") {
      this.handleToolCallRequest(requestId, p.payload);
      return;
    }
    const result = parseRequestPayload(p.type, p.payload);
    if (result.ok) {
      this.pushEvent?.(result.value);
    } else {
      this.emitParseError("UNKNOWN_REQUEST_TYPE", result.error);
    }
  }
  async handleToolCallRequest(requestId, request) {
    const handler = this.externalToolHandlers.get(request.name);
    let returnValue;
    if (!handler) {
      returnValue = {
        is_error: true,
        output: `Unknown external tool: ${request.name}`,
        message: `Tool "${request.name}" is not registered`,
        display: []
      };
    } else {
      try {
        const params = request.arguments ? JSON.parse(request.arguments) : {};
        const result = await handler(params);
        returnValue = {
          is_error: false,
          output: result.output,
          message: result.message,
          display: []
        };
      } catch (err) {
        const message = err instanceof Error ? err.message : String(err);
        returnValue = {
          is_error: true,
          output: message,
          message: `Tool execution failed: ${message}`,
          display: []
        };
      }
    }
    this.writeLine({
      jsonrpc: "2.0",
      id: requestId,
      result: {
        tool_call_id: request.id,
        return_value: returnValue
      }
    });
  }
  emitParseError(code, message, raw) {
    const error = { type: "error", code, message, raw: raw?.slice(0, 500) };
    this.pushEvent?.(error);
  }
  // Private: Process Lifecycle
  handleProcessError(err) {
    log.protocol("Process error: %s", err.message);
    const error = new TransportError("PROCESS_CRASHED", `CLI process error: ${err.message}`, err);
    for (const pending of this.pendingRequests.values()) {
      pending.reject(error);
    }
    this.finishEvents?.();
    this.cleanup();
  }
  handleProcessExit(code) {
    log.protocol("Process exited with code: %d", code);
    if (code !== 0 && code !== null) {
      const errorMsg = this.stderrBuffer.trim() ? `CLI exited with code ${code}: ${this.stderrBuffer.trim()}` : `CLI exited with code ${code}`;
      const error = new TransportError("PROCESS_CRASHED", errorMsg);
      for (const pending of this.pendingRequests.values()) {
        pending.reject(error);
      }
    }
    this.finishEvents?.();
    this.cleanup();
  }
  cleanup() {
    this.readline?.removeAllListeners();
    this.readline?.close();
    this.readline = null;
    this.process?.removeAllListeners();
    this.process?.stdout?.removeAllListeners();
    this.process?.stderr?.removeAllListeners();
    this.process = null;
    this.pushEvent = null;
    this.finishEvents = null;
    this.pendingRequests.clear();
    this.externalToolHandlers.clear();
  }
};

// session.ts
var TurnImpl = class {
  constructor(getClient, getCurrentClient, getNextPending, clearPending, onComplete) {
    this.getClient = getClient;
    this.getCurrentClient = getCurrentClient;
    this.getNextPending = getNextPending;
    this.clearPending = clearPending;
    this.onComplete = onComplete;
    let resolve;
    let reject;
    this.result = new Promise((res, rej) => {
      resolve = res;
      reject = rej;
    });
    this.result.catch(() => {
    });
    this.resolveResult = resolve;
    this.rejectResult = reject;
  }
  result;
  resolveResult;
  rejectResult;
  interrupted = false;
  async *[Symbol.asyncIterator]() {
    try {
      let result;
      let content;
      while (!this.interrupted && (content = this.getNextPending()) !== void 0) {
        result = yield* this.processOne(content);
      }
      this.onComplete();
      this.resolveResult(result);
      return result;
    } catch (err) {
      this.onComplete();
      this.rejectResult(err);
      throw err;
    }
  }
  async *processOne(content) {
    const client = await this.getClient();
    const stream = client.sendPrompt(content);
    for await (const event of stream.events) {
      yield event;
    }
    return await stream.result;
  }
  async interrupt() {
    this.interrupted = true;
    this.clearPending();
    const client = this.getCurrentClient();
    if (client?.isRunning) {
      return client.sendCancel();
    }
  }
  async approve(requestId, response) {
    const client = this.getCurrentClient();
    if (!client?.isRunning) {
      throw new SessionError("SESSION_CLOSED", "Cannot approve: no active client");
    }
    return client.sendApproval(requestId, response);
  }
  async respondQuestion(rpcRequestId, questionRequestId, answers) {
    const client = this.getCurrentClient();
    if (!client?.isRunning) {
      throw new SessionError("SESSION_CLOSED", "Cannot respond to question: no active client");
    }
    return client.sendQuestionResponse(rpcRequestId, questionRequestId, answers);
  }
};
var SessionImpl = class {
  _sessionId;
  _workDir;
  _clientInfo;
  _model;
  _thinking;
  _yoloMode;
  _executable;
  _env;
  _externalTools;
  _agentFile;
  _skillsDir;
  _shareDir;
  _slashCommands = [];
  _state = "idle";
  client = null;
  activeConfig = null;
  currentTurn = null;
  pendingMessages = [];
  constructor(options) {
    this._sessionId = options.sessionId ?? crypto.randomUUID();
    this._workDir = options.workDir;
    this._model = options.model;
    this._thinking = options.thinking ?? false;
    this._yoloMode = options.yoloMode ?? false;
    this._executable = options.executable ?? "kimi";
    this._env = options.env ?? {};
    this._externalTools = options.externalTools ?? [];
    this._agentFile = options.agentFile;
    this._skillsDir = options.skillsDir;
    this._shareDir = options.shareDir;
    this._clientInfo = options.clientInfo;
    log.session("Created session %s in %s", this._sessionId, this._workDir);
  }
  get sessionId() {
    return this._sessionId;
  }
  get workDir() {
    return this._workDir;
  }
  get state() {
    return this._state;
  }
  get slashCommands() {
    return this._slashCommands;
  }
  get model() {
    return this._model;
  }
  set model(v) {
    this._model = v;
  }
  get thinking() {
    return this._thinking;
  }
  set thinking(v) {
    this._thinking = v;
  }
  get yoloMode() {
    return this._yoloMode;
  }
  set yoloMode(v) {
    this._yoloMode = v;
  }
  get executable() {
    return this._executable;
  }
  set executable(v) {
    this._executable = v;
  }
  get env() {
    return this._env;
  }
  set env(v) {
    this._env = v;
  }
  get externalTools() {
    return this._externalTools;
  }
  set externalTools(v) {
    this._externalTools = v;
  }
  prompt(content) {
    if (this._state === "closed") {
      throw new SessionError("SESSION_CLOSED", "Session is closed");
    }
    this.pendingMessages.push(content);
    log.session("Queued prompt, pending: %d", this.pendingMessages.length);
    if (this._state === "active" && this.currentTurn) {
      return this.currentTurn;
    }
    this._state = "active";
    this.currentTurn = new TurnImpl(
      () => this.getClientWithConfigCheck(),
      () => this.client,
      () => this.pendingMessages.shift(),
      () => {
        this.pendingMessages = [];
      },
      () => {
        if (this._state === "active") {
          this._state = "idle";
        }
        this.currentTurn = null;
        log.session("Turn completed, state: %s", this._state);
      }
    );
    return this.currentTurn;
  }
  async close() {
    if (this._state === "closed") {
      return;
    }
    log.session("Closing session %s", this._sessionId);
    this._state = "closed";
    this.currentTurn = null;
    this.pendingMessages = [];
    if (this.client) {
      try {
        await this.client.stop();
      } catch (err) {
        log.session("Error during close: %O", err);
      }
      this.client = null;
      this.activeConfig = null;
    }
  }
  [Symbol.asyncDispose]() {
    return this.close();
  }
  async getClientWithConfigCheck() {
    const currentConfig = this.snapshotConfig();
    if (this.client?.isRunning && this.activeConfig && !this.configChanged(currentConfig)) {
      return this.client;
    }
    if (this.client) {
      log.session("Config changed, restarting client");
      await this.client.stop();
      this.client = null;
    }
    this.client = new ProtocolClient();
    const envVars = this._shareDir ? { KIMI_SHARE_DIR: this._shareDir, ...this._env } : this._env;
    const initResult = await this.client.start({
      sessionId: this._sessionId,
      workDir: this._workDir,
      model: this._model,
      thinking: this._thinking,
      yoloMode: this._yoloMode,
      executablePath: this._executable,
      environmentVariables: envVars,
      externalTools: this._externalTools,
      agentFile: this._agentFile,
      skillsDir: this._skillsDir,
      clientInfo: this._clientInfo
    });
    this._slashCommands = initResult.slash_commands;
    this.activeConfig = currentConfig;
    return this.client;
  }
  snapshotConfig() {
    return {
      model: this._model,
      thinking: this._thinking,
      yoloMode: this._yoloMode,
      executable: this._executable,
      env: JSON.stringify(this._env),
      externalTools: JSON.stringify(this._externalTools.map((t) => t.name)),
      skillsDir: this._skillsDir
    };
  }
  configChanged(current) {
    const active = this.activeConfig;
    return current.model !== active.model || current.thinking !== active.thinking || current.yoloMode !== active.yoloMode || current.executable !== active.executable || current.env !== active.env || current.externalTools !== active.externalTools || current.skillsDir !== active.skillsDir;
  }
};
function createSession(options) {
  return new SessionImpl(options);
}
async function prompt(content, options) {
  const session = createSession(options);
  try {
    const turn = session.prompt(content);
    const events = [];
    for await (const event of turn) {
      events.push(event);
    }
    return { result: await turn.result, events };
  } finally {
    await session.close();
  }
}
function hashPath(workDir) {
  return crypto.createHash("md5").update(workDir, "utf-8").digest("hex");
}
function createKimiPaths(shareDir) {
  const home = shareDir || process.env.KIMI_SHARE_DIR || path2.join(os.homedir(), ".kimi");
  return {
    home,
    config: path2.join(home, "config.toml"),
    mcpConfig: path2.join(home, "mcp.json"),
    sessionsDir(workDir) {
      return path2.join(home, "sessions", hashPath(workDir));
    },
    sessionDir(workDir, sessionId) {
      return path2.join(home, "sessions", hashPath(workDir), sessionId);
    },
    baselineDir(workDir, sessionId) {
      return path2.join(home, "sessions", hashPath(workDir), sessionId, "baseline");
    }
  };
}
var KimiPaths = createKimiPaths();

// storage.ts
function readKimiJson() {
  try {
    return JSON.parse(fs.readFileSync(path2.join(KimiPaths.home, "kimi.json"), "utf-8"));
  } catch {
    return null;
  }
}
function readSessionMetadata(sessionDir) {
  try {
    return JSON.parse(fs.readFileSync(path2.join(sessionDir, "metadata.json"), "utf-8"));
  } catch {
    return null;
  }
}
var UUID_REGEX = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
function getRegisteredWorkDirs(workspaceRoot) {
  const config = readKimiJson();
  if (!config?.work_dirs) {
    return [workspaceRoot];
  }
  const sep2 = path2.sep;
  const dirs = config.work_dirs.map((e) => e.path).filter((p) => p === workspaceRoot || p.startsWith(workspaceRoot + sep2));
  return dirs.length > 0 ? dirs : [workspaceRoot];
}
async function listSessionsForWorkspace(workspaceRoot) {
  const workDirs = getRegisteredWorkDirs(workspaceRoot);
  const allSessions = [];
  for (const workDir of workDirs) {
    const sessions = await listSessions(workDir);
    allSessions.push(...sessions);
  }
  return allSessions.sort((a, b) => b.updatedAt - a.updatedAt);
}
async function listSessions(workDir) {
  const sessionsDir = KimiPaths.sessionsDir(workDir);
  try {
    await fsp.access(sessionsDir);
  } catch {
    return [];
  }
  let entries;
  try {
    entries = await fsp.readdir(sessionsDir, { withFileTypes: true });
  } catch (err) {
    console.warn("[storage] Failed to read sessions:", err);
    return [];
  }
  const sessions = [];
  for (const entry of entries) {
    if (!entry.isDirectory() || !UUID_REGEX.test(entry.name)) {
      continue;
    }
    const sessionId = entry.name;
    const sessionDir = path2.join(sessionsDir, sessionId);
    const wireFile = path2.join(sessionDir, "wire.jsonl");
    if (!fs.existsSync(wireFile)) {
      continue;
    }
    try {
      const stat3 = await fsp.stat(wireFile);
      if (stat3.size === 0) {
        continue;
      }
      const metadata = readSessionMetadata(sessionDir);
      const brief = metadata?.title || await getFirstUserMessage(wireFile);
      if (!brief) {
        continue;
      }
      sessions.push({
        id: sessionId,
        workDir,
        contextFile: wireFile,
        updatedAt: stat3.mtimeMs,
        brief
      });
    } catch (err) {
      log.storage("Failed to stat session %s: %O", sessionId, err);
    }
  }
  return sessions.sort((a, b) => b.updatedAt - a.updatedAt);
}
async function deleteSession(workDir, sessionId) {
  const sessionDir = path2.join(KimiPaths.sessionsDir(workDir), sessionId);
  try {
    await fsp.access(sessionDir);
  } catch {
    return false;
  }
  try {
    await fsp.rm(sessionDir, { recursive: true, force: true });
    log.storage("Deleted session %s", sessionId);
    return true;
  } catch (err) {
    log.storage("Failed to delete session %s: %O", sessionId, err);
    return false;
  }
}
async function getFirstUserMessage(wireFile) {
  try {
    const stream = fs.createReadStream(wireFile, { encoding: "utf-8" });
    const rl = readline.createInterface({ input: stream, crlfDelay: Infinity });
    for await (const line of rl) {
      if (!line.trim()) {
        continue;
      }
      try {
        const record = JSON.parse(line);
        if (record.message?.type !== "TurnBegin") {
          continue;
        }
        const userInput = record.message.payload?.user_input;
        const text = extractUserText(userInput);
        if (text) {
          rl.close();
          stream.destroy();
          return text;
        }
      } catch {
        continue;
      }
    }
  } catch (err) {
    log.storage("Failed to read wire file: %O", err);
  }
  return "";
}
function extractUserText(userInput) {
  if (typeof userInput === "string") {
    return stripFileTags(userInput);
  }
  if (Array.isArray(userInput)) {
    const textParts = userInput.filter((p) => p.type === "text").map((p) => p.text);
    return stripFileTags(textParts.join("\n"));
  }
  return "";
}
function stripFileTags(text) {
  return text.replace(/<uploaded_files>[\s\S]*?<\/uploaded_files>\s*/g, "").replace(/<document[^>]*>[\s\S]*?<\/document>\s*/g, "").replace(/<image[^>]*>[\s\S]*?<\/image>\s*/g, "").trim();
}
async function forkSession(options) {
  const { workDir, sourceSessionId, turnIndex } = options;
  if (!UUID_REGEX.test(sourceSessionId)) {
    throw new Error(`Invalid session ID: ${sourceSessionId}`);
  }
  if (turnIndex < 0) {
    throw new Error(`Invalid turn index: ${turnIndex}`);
  }
  const sourceDir = KimiPaths.sessionDir(workDir, sourceSessionId);
  const sourceWireFile = path2.join(sourceDir, "wire.jsonl");
  const sourceContextFile = path2.join(sourceDir, "context.jsonl");
  try {
    await fsp.access(sourceWireFile);
  } catch {
    throw new Error(`Source session not found: ${sourceSessionId}`);
  }
  const newSessionId = crypto.randomUUID();
  const newSessionDir = KimiPaths.sessionDir(workDir, newSessionId);
  await fsp.mkdir(newSessionDir, { recursive: true });
  const wireLines = await truncateWireAtTurn(sourceWireFile, turnIndex);
  if (wireLines.length === 0) {
    await fsp.rm(newSessionDir, { recursive: true, force: true });
    throw new Error(`Turn ${turnIndex} not found in session`);
  }
  await fsp.writeFile(path2.join(newSessionDir, "wire.jsonl"), wireLines.join("\n") + "\n");
  try {
    await fsp.access(sourceContextFile);
    const contextLines = await truncateContextAtTurn(sourceContextFile, turnIndex);
    if (contextLines.length > 0) {
      await fsp.writeFile(path2.join(newSessionDir, "context.jsonl"), contextLines.join("\n") + "\n");
    }
  } catch {
  }
  log.storage("Forked session %s -> %s at turn %d", sourceSessionId, newSessionId, turnIndex);
  return {
    sessionId: newSessionId,
    sessionDir: newSessionDir
  };
}
async function truncateWireAtTurn(wireFile, turnIndex) {
  const lines = [];
  let turnCount = 0;
  let lastTurnEndIndex = -1;
  const stream = fs.createReadStream(wireFile, { encoding: "utf-8" });
  const rl = readline.createInterface({ input: stream, crlfDelay: Infinity });
  for await (const line of rl) {
    if (!line.trim()) {
      continue;
    }
    try {
      const record = JSON.parse(line);
      const messageType = record.message?.type;
      if (messageType === "TurnBegin") {
        if (turnCount > turnIndex) {
          break;
        }
        turnCount++;
      }
      lines.push(line);
      if (messageType === "TurnEnd" && turnCount === turnIndex + 1) {
        lastTurnEndIndex = lines.length;
        break;
      }
    } catch {
      if (turnCount > 0 && turnCount <= turnIndex + 1) {
        lines.push(line);
      }
    }
  }
  rl.close();
  stream.destroy();
  if (lastTurnEndIndex > 0) {
    return lines.slice(0, lastTurnEndIndex);
  }
  for (let i = lines.length - 1; i >= 0; i--) {
    try {
      const record = JSON.parse(lines[i]);
      if (record.message?.type === "TurnEnd") {
        log.storage("Target turn incomplete, truncating to last complete turn");
        return lines.slice(0, i + 1);
      }
    } catch {
      continue;
    }
  }
  log.storage("No complete turn found in session");
  return [];
}
async function truncateContextAtTurn(contextFile, turnIndex) {
  const lines = [];
  let userMessageCount = 0;
  const stream = fs.createReadStream(contextFile, { encoding: "utf-8" });
  const rl = readline.createInterface({ input: stream, crlfDelay: Infinity });
  for await (const line of rl) {
    if (!line.trim()) {
      continue;
    }
    try {
      const record = JSON.parse(line);
      const role = record.role;
      if (role === "_checkpoint" || role === "_usage") {
        lines.push(line);
        continue;
      }
      if (role === "user") {
        if (userMessageCount > turnIndex) {
          break;
        }
        userMessageCount++;
      } else if (role === "assistant" && userMessageCount > turnIndex + 1) {
        break;
      }
      lines.push(line);
    } catch {
      lines.push(line);
    }
  }
  rl.close();
  stream.destroy();
  return removeIncompleteToolCalls(lines);
}
function removeIncompleteToolCalls(lines) {
  const expectedIds = /* @__PURE__ */ new Set();
  const receivedIds = /* @__PURE__ */ new Set();
  const assistantIndicesWithToolCalls = [];
  for (let i = 0; i < lines.length; i++) {
    try {
      const record = JSON.parse(lines[i]);
      if (record.role === "assistant" && Array.isArray(record.tool_calls)) {
        for (const tc of record.tool_calls) {
          if (tc.id) expectedIds.add(tc.id);
        }
        if (record.tool_calls.length > 0) {
          assistantIndicesWithToolCalls.push(i);
        }
      } else if (record.role === "tool" && record.tool_call_id) {
        receivedIds.add(record.tool_call_id);
      }
    } catch {
      continue;
    }
  }
  const hasIncomplete = [...expectedIds].some((id) => !receivedIds.has(id));
  if (!hasIncomplete) {
    return lines;
  }
  for (let i = assistantIndicesWithToolCalls.length - 1; i >= 0; i--) {
    const idx = assistantIndicesWithToolCalls[i];
    try {
      const record = JSON.parse(lines[idx]);
      const ids = (record.tool_calls || []).map((tc) => tc.id).filter(Boolean);
      const incomplete = ids.some((id) => !receivedIds.has(id));
      if (incomplete) {
        log.storage("Removing incomplete assistant message at index %d", idx);
        const before = lines.slice(0, idx);
        const after = lines.slice(idx + 1).filter((l) => {
          try {
            const r = JSON.parse(l);
            return r.role === "_checkpoint" || r.role === "_usage";
          } catch {
            return false;
          }
        });
        return [...before, ...after];
      }
    } catch {
      continue;
    }
  }
  return lines;
}
async function parseSessionEvents(workDir, sessionId) {
  const sessionDir = KimiPaths.sessionDir(workDir, sessionId);
  const wireFile = path2.join(sessionDir, "wire.jsonl");
  if (!fs.existsSync(wireFile)) {
    log.history("Wire file not found for session: %s", sessionId);
    return [];
  }
  const stat3 = await fsp.stat(wireFile);
  if (stat3.size === 0) {
    log.history("Wire file is empty for session: %s", sessionId);
    return [];
  }
  log.history("Parsing wire file: %s (%d bytes)", wireFile, stat3.size);
  return parseWireFile(wireFile);
}
async function parseWireFile(filePath) {
  const events = [];
  const stream = fs.createReadStream(filePath, { encoding: "utf-8" });
  const rl = readline.createInterface({ input: stream, crlfDelay: Infinity });
  for await (const line of rl) {
    if (!line.trim()) {
      continue;
    }
    try {
      const record = JSON.parse(line);
      const event = parseWireRecord(record);
      if (event) {
        events.push(event);
      }
    } catch {
    }
  }
  log.history("Parsed %d events from wire file", events.length);
  return events;
}
function parseWireRecord(record) {
  if (!record || typeof record !== "object") {
    return null;
  }
  const rec = record;
  const message = rec.message;
  if (!message?.type) {
    return null;
  }
  const result = parseEventPayload(message.type, message.payload);
  return result.ok ? result.value : null;
}
var OAuthConfigSchema = z.object({
  storage: z.string(),
  key: z.string()
});
var ProviderTypeSchema = z.enum(["kimi", "openai_legacy", "openai_responses", "anthropic", "google_genai", "gemini", "vertexai"]);
var LLMProviderSchema = z.object({
  type: ProviderTypeSchema,
  base_url: z.string(),
  api_key: z.string(),
  env: z.record(z.string(), z.string()).optional(),
  custom_headers: z.record(z.string(), z.string()).optional(),
  oauth: OAuthConfigSchema.optional()
});
var ModelCapabilitySchema = z.enum(["thinking", "always_thinking", "image_in", "video_in"]);
var LLMModelSchema = z.object({
  provider: z.string(),
  model: z.string(),
  max_context_size: z.number().int().positive(),
  capabilities: z.array(ModelCapabilitySchema).optional()
});
var LoopControlSchema = z.object({
  max_steps_per_turn: z.number().int().min(1).default(100),
  max_retries_per_step: z.number().int().min(1).default(3),
  max_ralph_iterations: z.number().int().min(-1).default(0)
});
var MoonshotSearchConfigSchema = z.object({
  base_url: z.string(),
  api_key: z.string(),
  custom_headers: z.record(z.string(), z.string()).optional()
});
var MoonshotFetchConfigSchema = z.object({
  base_url: z.string(),
  api_key: z.string(),
  custom_headers: z.record(z.string(), z.string()).optional()
});
var ServicesSchema = z.object({
  moonshot_search: MoonshotSearchConfigSchema.optional(),
  moonshot_fetch: MoonshotFetchConfigSchema.optional()
});
var MCPClientConfigSchema = z.object({
  tool_call_timeout_ms: z.number().int().positive().default(6e4)
});
var MCPConfigSchema = z.object({
  client: MCPClientConfigSchema.default({})
});
var DefaultThinkingSchema = z.preprocess((val) => {
  if (val === "on") {
    return true;
  }
  if (val === "off") {
    return false;
  }
  return val;
}, z.boolean()).default(false);
var ConfigSchema = z.object({
  default_model: z.string().default(""),
  default_thinking: DefaultThinkingSchema,
  models: z.record(z.string(), LLMModelSchema).default({}),
  providers: z.record(z.string(), LLMProviderSchema).default({}),
  loop_control: LoopControlSchema.default({}),
  services: ServicesSchema.default({}),
  mcp: MCPConfigSchema.default({})
});
function readConfigToml(configPath) {
  if (!fs.existsSync(configPath)) {
    return null;
  }
  try {
    return toml.parse(fs.readFileSync(configPath, "utf-8"));
  } catch (err) {
    log.config("Failed to read/parse config.toml: %O", err);
    return null;
  }
}
function parseConfig(shareDir) {
  const paths = shareDir ? createKimiPaths(shareDir) : KimiPaths;
  const raw = readConfigToml(paths.config);
  if (!raw) {
    log.config("Config file not found: %s", paths.config);
    return { defaultModel: null, defaultThinking: false, models: [] };
  }
  const parsed = ConfigSchema.safeParse(raw);
  if (!parsed.success) {
    log.config("Failed to validate config.toml: %O", parsed.error);
    return { defaultModel: null, defaultThinking: false, models: [] };
  }
  const config = parsed.data;
  log.config("Parsed config with %d models", Object.keys(config.models).length);
  return toKimiConfig(config);
}
function toKimiConfig(config) {
  const models = Object.entries(config.models).map(([id, model]) => ({
    id,
    name: id,
    capabilities: model.capabilities ?? []
  }));
  models.sort((a, b) => a.name.localeCompare(b.name));
  return {
    defaultModel: config.default_model || null,
    defaultThinking: config.default_thinking,
    models
  };
}
function isPlainObject(val) {
  return !!val && typeof val === "object" && !Array.isArray(val);
}
function isLoggedIn(shareDir) {
  const paths = shareDir ? createKimiPaths(shareDir) : KimiPaths;
  const raw = readConfigToml(paths.config);
  if (!raw || !isPlainObject(raw)) {
    return false;
  }
  const providers = raw["providers"];
  if (!isPlainObject(providers)) {
    return false;
  }
  const provider = providers["managed:kimi-code"];
  if (!isPlainObject(provider)) {
    return false;
  }
  const apiKey = provider["api_key"];
  const hasApiKey = typeof apiKey === "string" && apiKey.length > 0;
  const oauth = provider["oauth"];
  const hasOAuth = isPlainObject(oauth);
  return hasApiKey || hasOAuth;
}
function saveDefaultModel(modelId, thinking, shareDir) {
  const paths = shareDir ? createKimiPaths(shareDir) : KimiPaths;
  const configPath = paths.config;
  if (!fs.existsSync(configPath)) {
    let content2 = `default_model = "${modelId}"
`;
    if (thinking !== void 0) {
      content2 += `default_thinking = ${thinking}
`;
    }
    fs.writeFileSync(configPath, content2, "utf-8");
    log.config("Created config with default model: %s", modelId);
    return;
  }
  let content = fs.readFileSync(configPath, "utf-8");
  const modelRegex = /^default_model\s*=\s*"[^"]*"/m;
  if (modelRegex.test(content)) {
    content = content.replace(modelRegex, `default_model = "${modelId}"`);
  } else {
    content = `default_model = "${modelId}"
` + content;
  }
  if (thinking !== void 0) {
    const thinkingRegex = /^default_thinking\s*=\s*(?:true|false|"[^"]*")/m;
    const thinkingValue = thinking ? "true" : "false";
    if (thinkingRegex.test(content)) {
      content = content.replace(thinkingRegex, `default_thinking = ${thinkingValue}`);
    } else {
      content = content.replace(/^(default_model\s*=\s*"[^"]*")/m, `$1
default_thinking = ${thinkingValue}`);
    }
  }
  fs.writeFileSync(configPath, content, "utf-8");
  log.config("Updated default model: %s, thinking: %s", modelId, thinking);
}
function getModelById(models, modelId) {
  return models.find((m) => m.id === modelId);
}
function getModelThinkingMode(model) {
  if (model.name.toLowerCase().includes("think")) {
    return "always";
  }
  if (model.capabilities.includes("always_thinking")) {
    return "always";
  }
  if (model.capabilities.includes("thinking")) {
    return "switch";
  }
  return "none";
}
function isModelThinking(models, modelId) {
  const model = getModelById(models, modelId);
  if (!model) {
    return false;
  }
  const mode = getModelThinkingMode(model);
  return mode === "always" || mode === "switch";
}
var DEFAULT_EXECUTABLE = "kimi";
var handleError = (err) => ({
  success: false,
  error: err instanceof Error ? err.message : String(err)
});
async function authMCP(serverName, options) {
  const executable = options?.executable ?? DEFAULT_EXECUTABLE;
  log.cli("Running MCP auth for: %s", serverName);
  await runCliCommand(executable, ["mcp", "auth", serverName], { env: options?.env });
}
async function resetAuthMCP(serverName, options) {
  const executable = options?.executable ?? DEFAULT_EXECUTABLE;
  log.cli("Running MCP reset-auth for: %s", serverName);
  await runCliCommand(executable, ["mcp", "reset-auth", serverName], { env: options?.env });
}
function testMCP(serverName, options) {
  const executable = options?.executable ?? DEFAULT_EXECUTABLE;
  log.cli("Running MCP test for: %s", serverName);
  return runCliCommand(executable, ["mcp", "test", serverName], { env: options?.env }).then((output) => ({ success: true, output })).catch((err) => ({
    success: false,
    output: err instanceof Error ? err.message : String(err)
  }));
}
function login(options) {
  const executable = options?.executable ?? DEFAULT_EXECUTABLE;
  log.cli("Running login --json");
  return runCliCommand(executable, ["login", "--json"], {
    env: options?.env,
    onLine: (line) => {
      try {
        const msg = JSON.parse(line);
        if (msg.type === "verification_url" && msg.data?.verification_url) {
          options?.onUrl?.(msg.data.verification_url);
        }
      } catch {
      }
    }
  }).then(() => ({ success: true })).catch(handleError);
}
function logout(options) {
  const executable = options?.executable ?? DEFAULT_EXECUTABLE;
  log.cli("Running logout");
  return runCliCommand(executable, ["logout"], { env: options?.env }).then(() => ({ success: true })).catch(handleError);
}
function runCliCommand(executable, args, options = {}) {
  return new Promise((resolve, reject) => {
    log.cli("Executing: %s %o", executable, args);
    const proc = spawn(executable, args, {
      stdio: ["ignore", "pipe", "pipe"],
      env: { ...process.env, ...options.env, NO_COLOR: "1" }
    });
    let stdout = "";
    let stderr = "";
    const rl = createInterface({ input: proc.stdout, terminal: false });
    rl.on("line", (line) => {
      stdout += line + "\n";
      if (line.trim() && options.onLine) {
        log.cli("cli stdout: %s", line);
        options.onLine(line);
      }
    });
    proc.stderr?.on("data", (data) => {
      const chunk = data.toString();
      stderr += chunk;
      if (chunk.trim()) {
        log.cli("cli stderr: %s", chunk.trim());
      }
    });
    proc.on("error", (err) => {
      log.cli("Command error: %O", err);
      reject(new TransportError("CLI_NOT_FOUND", `Failed to run CLI: ${err.message}`, err));
    });
    proc.on("close", (code) => {
      log.cli("Command exited with code: %d", code);
      if (code === 0) {
        resolve(stdout.trim());
      } else {
        const errorMessage = stderr.trim() || stdout.trim() || `CLI exited with code ${code}`;
        reject(new CliError("UNKNOWN", errorMessage, code ?? void 0));
      }
    });
  });
}
function toJsonSchema(schema) {
  if ("toJSONSchema" in z && typeof z.toJSONSchema === "function") {
    return z.toJSONSchema(schema);
  }
  return zodToJsonSchema(schema, { $refStrategy: "none" });
}
function createExternalTool(definition) {
  const jsonSchema = toJsonSchema(definition.parameters);
  const handler = async (params) => {
    const parsed = definition.parameters.parse(params);
    return definition.handler(parsed);
  };
  return {
    name: definition.name,
    description: definition.description,
    parameters: jsonSchema,
    handler
  };
}

// utils.ts
function extractBrief(display) {
  const brief = display?.find((d) => d.type === "brief");
  return brief?.text ?? "";
}
function collectText(events) {
  return events.filter((e) => e.type === "ContentPart" && "payload" in e && e.payload.type === "text").map((e) => e.payload.text).join("");
}
function extractTextFromContentParts(parts) {
  return parts.filter((p) => p.type === "text").map((p) => p.text).join("\n");
}
function formatContentOutput(output) {
  if (typeof output === "string") {
    return output;
  }
  if (!Array.isArray(output)) {
    return JSON.stringify(output);
  }
  return output.map((item) => {
    if (typeof item === "string") {
      return item;
    }
    if (item.type === "text") {
      return item.text;
    }
    return `[${item.type}]`;
  }).filter(Boolean).join("\n");
}

export { AgentSdkError, CliError, CliErrorCodes, ContentPartSchema, DisplayBlockSchema, InitializeResultSchema, KimiPaths, ProtocolClient, ProtocolError, ProtocolErrorCodes, ReplayResultSchema, RunResultSchema, SessionError, SessionErrorCodes, SlashCommandInfoSchema, ToolCallSchema, ToolResultSchema, TransportError, TransportErrorCodes, authMCP, collectText, createExternalTool, createKimiPaths, createSession, deleteSession, disableLogs, enableLogs, extractBrief, extractTextFromContentParts, forkSession, formatContentOutput, getErrorCategory, getErrorCode, getModelById, getModelThinkingMode, getRegisteredWorkDirs, isAgentSdkError, isLoggedIn, isModelThinking, listSessions, listSessionsForWorkspace, login, logout, parseConfig, parseEventPayload, parseRequestPayload, parseSessionEvents, prompt, resetAuthMCP, saveDefaultModel, setLogSink, testMCP };
