// @moonshot-ai/kimi-agent-sdk/dist/schema.cjs
'use strict';

var zod = require('zod');

// schema.ts
var ApprovalResponseSchema = zod.z.enum(["approve", "approve_for_session", "reject"]);
var ContentPartSchema = zod.z.discriminatedUnion("type", [
  zod.z.object({
    // Text type
    type: zod.z.literal("text"),
    // Text content
    text: zod.z.string()
  }),
  zod.z.object({
    // Think type, appears only in thinking mode
    type: zod.z.literal("think"),
    // Think content
    think: zod.z.string(),
    // Encrypted think content or signature
    encrypted: zod.z.string().nullable().optional()
  }),
  zod.z.object({
    // Image type
    type: zod.z.literal("image_url"),
    image_url: zod.z.object({
      // Image URL, usually a data URI (e.g., data:image/png;base64,...)
      url: zod.z.string(),
      // Image ID, used to distinguish different images
      id: zod.z.string().nullable().optional()
    })
  }),
  zod.z.object({
    // Audio type
    type: zod.z.literal("audio_url"),
    audio_url: zod.z.object({
      // Audio URL, usually a data URI (e.g., data:audio/aac;base64,...)
      url: zod.z.string(),
      // Audio ID, used to distinguish different audio
      id: zod.z.string().nullable().optional()
    })
  }),
  zod.z.object({
    // Video type
    type: zod.z.literal("video_url"),
    video_url: zod.z.object({
      // Video URL, usually a data URI (e.g., data:video/mp4;base64,...)
      url: zod.z.string(),
      // Video ID, used to distinguish different videos
      id: zod.z.string().nullable().optional()
    })
  })
]);
var TokenUsageSchema = zod.z.object({
  // Number of input tokens (excluding cache)
  input_other: zod.z.number(),
  // Number of output tokens
  output: zod.z.number(),
  // Number of input tokens read from cache
  input_cache_read: zod.z.number(),
  // Number of input tokens written to cache
  input_cache_creation: zod.z.number()
});
var BriefBlockSchema = zod.z.object({
  type: zod.z.literal("brief"),
  // Brief text content
  text: zod.z.string()
});
var DiffBlockSchema = zod.z.object({
  type: zod.z.literal("diff"),
  // File path
  path: zod.z.string(),
  // Original content
  old_text: zod.z.string(),
  // New content
  new_text: zod.z.string()
});
var TodoBlockSchema = zod.z.object({
  type: zod.z.literal("todo"),
  // Todo list
  items: zod.z.array(
    zod.z.object({
      // Todo item title
      title: zod.z.string(),
      // Status: pending | in_progress | done
      status: zod.z.enum(["pending", "in_progress", "done"])
    })
  )
});
var ShellBlockSchema = zod.z.object({
  type: zod.z.literal("shell"),
  language: zod.z.string(),
  command: zod.z.string()
});
var RawDisplayBlockSchema = zod.z.object({
  type: zod.z.string(),
  text: zod.z.string().optional(),
  path: zod.z.string().optional(),
  old_text: zod.z.string().optional(),
  new_text: zod.z.string().optional(),
  items: zod.z.array(zod.z.object({ title: zod.z.string(), status: zod.z.enum(["pending", "in_progress", "done"]) })).optional(),
  language: zod.z.string().optional(),
  command: zod.z.string().optional()
}).passthrough();
var DisplayBlockSchema = RawDisplayBlockSchema.transform((raw) => {
  switch (raw.type) {
    case "brief":
      if (typeof raw.text === "string") {
        return { type: "brief", text: raw.text };
      }
      break;
    case "diff":
      if (typeof raw.path === "string" && typeof raw.old_text === "string" && typeof raw.new_text === "string") {
        return { type: "diff", path: raw.path, old_text: raw.old_text, new_text: raw.new_text };
      }
      break;
    case "todo":
      if (Array.isArray(raw.items)) {
        return { type: "todo", items: raw.items };
      }
      break;
    case "shell":
      if (typeof raw.language === "string" && typeof raw.command === "string") {
        return { type: "shell", language: raw.language, command: raw.command };
      }
      break;
  }
  const { type, ...rest } = raw;
  return { type, data: rest };
});
var ToolCallSchema = zod.z.object({
  // Always "function"
  type: zod.z.literal("function"),
  // Tool call ID, used to match with ToolResult
  id: zod.z.string(),
  function: zod.z.object({
    // Tool name, e.g. "Shell", "ReadFile", "WriteFile"
    name: zod.z.string(),
    // JSON-encoded arguments string, may be incomplete during streaming
    arguments: zod.z.string().nullable().optional()
  }),
  // Extra metadata
  extras: zod.z.record(zod.z.string(), zod.z.unknown()).nullable().optional()
});
var ToolCallPartSchema = zod.z.object({
  // Argument chunk, appended to the last ToolCall's arguments
  arguments_part: zod.z.string().nullable().optional()
});
var ToolReturnValueSchema = zod.z.object({
  is_error: zod.z.boolean(),
  output: zod.z.union([zod.z.string(), zod.z.array(ContentPartSchema)]),
  message: zod.z.string(),
  display: zod.z.array(DisplayBlockSchema),
  extras: zod.z.record(zod.z.string(), zod.z.unknown()).nullable().optional()
});
var ToolResultSchema = zod.z.object({
  // Corresponding tool call ID
  tool_call_id: zod.z.string(),
  return_value: ToolReturnValueSchema
});
var ClientInfoSchema = zod.z.object({
  name: zod.z.string(),
  version: zod.z.string().optional()
});
var ServerInfoSchema = zod.z.object({
  name: zod.z.string(),
  version: zod.z.string()
});
var SlashCommandInfoSchema = zod.z.object({
  name: zod.z.string(),
  description: zod.z.string(),
  aliases: zod.z.array(zod.z.string())
});
var ExternalToolDefinitionSchema = zod.z.object({
  name: zod.z.string(),
  description: zod.z.string(),
  parameters: zod.z.record(zod.z.string(), zod.z.unknown())
});
var ExternalToolsResultSchema = zod.z.object({
  accepted: zod.z.array(zod.z.string()),
  rejected: zod.z.array(zod.z.object({ name: zod.z.string(), reason: zod.z.string() }))
});
var ClientCapabilitiesSchema = zod.z.object({
  // Whether the client supports handling QuestionRequest messages
  supports_question: zod.z.boolean().optional()
});
var ServerCapabilitiesSchema = zod.z.object({
  // Whether the server supports sending QuestionRequest messages
  supports_question: zod.z.boolean().optional()
});
var InitializeParamsSchema = zod.z.object({
  protocol_version: zod.z.string(),
  client: ClientInfoSchema.optional(),
  external_tools: zod.z.array(ExternalToolDefinitionSchema).optional(),
  capabilities: ClientCapabilitiesSchema.optional()
});
var InitializeResultSchema = zod.z.object({
  protocol_version: zod.z.string(),
  server: ServerInfoSchema,
  slash_commands: zod.z.array(SlashCommandInfoSchema),
  external_tools: ExternalToolsResultSchema.optional(),
  capabilities: ServerCapabilitiesSchema.optional()
});
var ToolCallRequestSchema = zod.z.object({
  id: zod.z.string(),
  name: zod.z.string(),
  arguments: zod.z.string().nullable().optional()
});
var QuestionOptionSchema = zod.z.object({
  // Option label
  label: zod.z.string(),
  // Option description
  description: zod.z.string().optional()
});
var QuestionItemSchema = zod.z.object({
  // Question text
  question: zod.z.string(),
  // Short header label (max 12 chars)
  header: zod.z.string().optional(),
  // Options (2-4)
  options: zod.z.array(QuestionOptionSchema),
  // Whether multiple selection is allowed
  multi_select: zod.z.boolean().optional()
});
var QuestionRequestSchema = zod.z.object({
  // Request ID, used for response reference
  id: zod.z.string(),
  // Associated tool call ID
  tool_call_id: zod.z.string(),
  // Questions (1-4)
  questions: zod.z.array(QuestionItemSchema)
});
var QuestionResponseSchema = zod.z.object({
  // Corresponding request ID
  request_id: zod.z.string(),
  // Answer mapping: question text -> selected option label(s)
  answers: zod.z.record(zod.z.string(), zod.z.string())
});
var TurnBeginSchema = zod.z.object({
  // User input, can be plain text or array of content parts
  user_input: zod.z.union([zod.z.string(), zod.z.array(ContentPartSchema)])
});
var StepBeginSchema = zod.z.object({
  // Step number, starting from 1
  n: zod.z.number()
});
var EmptyPayloadSchema = zod.z.object({});
var StatusUpdateSchema = zod.z.object({
  // Context usage ratio, float between 0 and 1
  context_usage: zod.z.number().nullable().optional(),
  // Token usage stats for the current step
  token_usage: TokenUsageSchema.nullable().optional(),
  // Message ID for the current step
  message_id: zod.z.string().nullable().optional()
});
var ApprovalRequestPayloadSchema = zod.z.object({
  // Request ID, referenced when responding
  id: zod.z.string(),
  // Associated tool call ID
  tool_call_id: zod.z.string(),
  // Sender (tool name), e.g. "Shell", "WriteFile"
  sender: zod.z.string(),
  // Action description, e.g. "run shell command"
  action: zod.z.string(),
  // Detailed description, e.g. "Run command `rm -rf /`"
  description: zod.z.string(),
  // Display blocks shown to the user
  display: zod.z.array(DisplayBlockSchema).optional()
});
var ApprovalResponseEventSchema = zod.z.object({
  // Resolved approval request ID
  request_id: zod.z.string(),
  // Approval result
  response: ApprovalResponseSchema
});
var EventSchemas = {
  TurnBegin: TurnBeginSchema,
  TurnEnd: EmptyPayloadSchema,
  StepBegin: StepBeginSchema,
  StepInterrupted: EmptyPayloadSchema,
  CompactionBegin: EmptyPayloadSchema,
  CompactionEnd: EmptyPayloadSchema,
  StatusUpdate: StatusUpdateSchema,
  ContentPart: ContentPartSchema,
  ToolCall: ToolCallSchema,
  ToolCallPart: ToolCallPartSchema,
  ToolResult: ToolResultSchema,
  ApprovalResponse: ApprovalResponseEventSchema
};
var RequestSchemas = {
  ApprovalRequest: ApprovalRequestPayloadSchema,
  ToolCallRequest: ToolCallRequestSchema,
  QuestionRequest: QuestionRequestSchema
};
function parseEventPayload(type, payload) {
  const schema = EventSchemas[type];
  if (!schema) {
    return { ok: false, error: `Unknown event type: ${type}` };
  }
  const result = schema.safeParse(payload);
  if (!result.success) {
    return { ok: false, error: `Invalid payload for ${type}: ${result.error.message}` };
  }
  return { ok: true, value: { type, payload: result.data } };
}
function parseWireEvent(raw) {
  const result = parseEventPayload(raw.type, raw.payload);
  if (result.ok) {
    return result.value;
  }
  return {
    type: "ParseError",
    payload: {
      code: "SUBAGENT_PARSE_FAILED",
      message: result.error,
      rawType: raw.type
    }
  };
}
var SubagentEventSchema = zod.z.lazy(
  () => zod.z.object({
    task_tool_call_id: zod.z.string(),
    event: zod.z.object({ type: zod.z.string(), payload: zod.z.unknown() }).transform(parseWireEvent)
  })
);
EventSchemas.SubagentEvent = SubagentEventSchema;
var RunResultSchema = zod.z.object({
  /**
   * Completion status
   * - `finished`: Turn completed normally
   * - `cancelled`: Turn was cancelled via cancel()
   * - `max_steps_reached`: Reached maximum step limit
   */
  status: zod.z.enum(["finished", "cancelled", "max_steps_reached"]),
  // When status is max_steps_reached, returns the number of steps executed
  steps: zod.z.number().optional()
});
var ReplayResultSchema = zod.z.object({
  /**
   * Replay completion status
   * - `finished`: Replay completed normally
   * - `cancelled`: Replay was cancelled via cancel()
   */
  status: zod.z.enum(["finished", "cancelled"]),
  // Number of events replayed
  events: zod.z.number(),
  // Number of requests replayed
  requests: zod.z.number()
});
var RpcErrorSchema = zod.z.object({
  // Error code
  code: zod.z.number(),
  // Error message
  message: zod.z.string(),
  // Extra data
  data: zod.z.unknown().optional()
});
var RpcMessageSchema = zod.z.object({
  jsonrpc: zod.z.string().optional(),
  id: zod.z.string().optional(),
  method: zod.z.string().optional(),
  params: zod.z.unknown().optional(),
  result: zod.z.unknown().optional(),
  error: RpcErrorSchema.optional()
});
function parseRequestPayload(type, payload) {
  const schema = RequestSchemas[type];
  if (!schema) {
    return { ok: false, error: `Unknown request type: ${type}` };
  }
  const result = schema.safeParse(payload);
  if (!result.success) {
    return { ok: false, error: `Invalid payload for ${type}: ${result.error.message}` };
  }
  return { ok: true, value: { type, payload: result.data } };
}
var ContextRecordSchema = zod.z.object({
  role: zod.z.string().optional(),
  content: zod.z.unknown().optional(),
  tool_calls: zod.z.array(
    zod.z.object({
      id: zod.z.string().optional(),
      function: zod.z.object({
        name: zod.z.string().optional(),
        arguments: zod.z.union([zod.z.string(), zod.z.record(zod.z.string(), zod.z.unknown())]).optional()
      }).optional()
    })
  ).optional(),
  tool_call_id: zod.z.string().optional()
});

exports.ApprovalRequestPayloadSchema = ApprovalRequestPayloadSchema;
exports.ApprovalResponseEventSchema = ApprovalResponseEventSchema;
exports.ApprovalResponseSchema = ApprovalResponseSchema;
exports.BriefBlockSchema = BriefBlockSchema;
exports.ClientCapabilitiesSchema = ClientCapabilitiesSchema;
exports.ClientInfoSchema = ClientInfoSchema;
exports.ContentPartSchema = ContentPartSchema;
exports.ContextRecordSchema = ContextRecordSchema;
exports.DiffBlockSchema = DiffBlockSchema;
exports.DisplayBlockSchema = DisplayBlockSchema;
exports.EmptyPayloadSchema = EmptyPayloadSchema;
exports.EventSchemas = EventSchemas;
exports.ExternalToolDefinitionSchema = ExternalToolDefinitionSchema;
exports.ExternalToolsResultSchema = ExternalToolsResultSchema;
exports.InitializeParamsSchema = InitializeParamsSchema;
exports.InitializeResultSchema = InitializeResultSchema;
exports.QuestionItemSchema = QuestionItemSchema;
exports.QuestionOptionSchema = QuestionOptionSchema;
exports.QuestionRequestSchema = QuestionRequestSchema;
exports.QuestionResponseSchema = QuestionResponseSchema;
exports.ReplayResultSchema = ReplayResultSchema;
exports.RequestSchemas = RequestSchemas;
exports.RpcErrorSchema = RpcErrorSchema;
exports.RpcMessageSchema = RpcMessageSchema;
exports.RunResultSchema = RunResultSchema;
exports.ServerCapabilitiesSchema = ServerCapabilitiesSchema;
exports.ServerInfoSchema = ServerInfoSchema;
exports.ShellBlockSchema = ShellBlockSchema;
exports.SlashCommandInfoSchema = SlashCommandInfoSchema;
exports.StatusUpdateSchema = StatusUpdateSchema;
exports.StepBeginSchema = StepBeginSchema;
exports.SubagentEventSchema = SubagentEventSchema;
exports.TodoBlockSchema = TodoBlockSchema;
exports.TokenUsageSchema = TokenUsageSchema;
exports.ToolCallPartSchema = ToolCallPartSchema;
exports.ToolCallRequestSchema = ToolCallRequestSchema;
exports.ToolCallSchema = ToolCallSchema;
exports.ToolResultSchema = ToolResultSchema;
exports.ToolReturnValueSchema = ToolReturnValueSchema;
exports.TurnBeginSchema = TurnBeginSchema;
exports.parseEventPayload = parseEventPayload;
exports.parseRequestPayload = parseRequestPayload;
