// @moonshot-ai/kimi-agent-sdk/dist/schema.mjs
import { z } from 'zod';

// schema.ts
var ApprovalResponseSchema = z.enum(["approve", "approve_for_session", "reject"]);
var ContentPartSchema = z.discriminatedUnion("type", [
  z.object({
    // Text type
    type: z.literal("text"),
    // Text content
    text: z.string()
  }),
  z.object({
    // Think type, appears only in thinking mode
    type: z.literal("think"),
    // Think content
    think: z.string(),
    // Encrypted think content or signature
    encrypted: z.string().nullable().optional()
  }),
  z.object({
    // Image type
    type: z.literal("image_url"),
    image_url: z.object({
      // Image URL, usually a data URI (e.g., data:image/png;base64,...)
      url: z.string(),
      // Image ID, used to distinguish different images
      id: z.string().nullable().optional()
    })
  }),
  z.object({
    // Audio type
    type: z.literal("audio_url"),
    audio_url: z.object({
      // Audio URL, usually a data URI (e.g., data:audio/aac;base64,...)
      url: z.string(),
      // Audio ID, used to distinguish different audio
      id: z.string().nullable().optional()
    })
  }),
  z.object({
    // Video type
    type: z.literal("video_url"),
    video_url: z.object({
      // Video URL, usually a data URI (e.g., data:video/mp4;base64,...)
      url: z.string(),
      // Video ID, used to distinguish different videos
      id: z.string().nullable().optional()
    })
  })
]);
var TokenUsageSchema = z.object({
  // Number of input tokens (excluding cache)
  input_other: z.number(),
  // Number of output tokens
  output: z.number(),
  // Number of input tokens read from cache
  input_cache_read: z.number(),
  // Number of input tokens written to cache
  input_cache_creation: z.number()
});
var BriefBlockSchema = z.object({
  type: z.literal("brief"),
  // Brief text content
  text: z.string()
});
var DiffBlockSchema = z.object({
  type: z.literal("diff"),
  // File path
  path: z.string(),
  // Original content
  old_text: z.string(),
  // New content
  new_text: z.string()
});
var TodoBlockSchema = z.object({
  type: z.literal("todo"),
  // Todo list
  items: z.array(
    z.object({
      // Todo item title
      title: z.string(),
      // Status: pending | in_progress | done
      status: z.enum(["pending", "in_progress", "done"])
    })
  )
});
var ShellBlockSchema = z.object({
  type: z.literal("shell"),
  language: z.string(),
  command: z.string()
});
var RawDisplayBlockSchema = z.object({
  type: z.string(),
  text: z.string().optional(),
  path: z.string().optional(),
  old_text: z.string().optional(),
  new_text: z.string().optional(),
  items: z.array(z.object({ title: z.string(), status: z.enum(["pending", "in_progress", "done"]) })).optional(),
  language: z.string().optional(),
  command: z.string().optional()
}).passthrough();
var DisplayBlockSchema = RawDisplayBlockSchema.transform((raw) => {
  switch (raw.type) {
    case "brief":
      if (typeof raw.text === "string") {
        return { type: "brief", text: raw.text };
      }
      break;
    case "diff":
      if (typeof raw.path === "string" && typeof raw.old_text === "string" && typeof raw.new_text === "string") {
        return { type: "diff", path: raw.path, old_text: raw.old_text, new_text: raw.new_text };
      }
      break;
    case "todo":
      if (Array.isArray(raw.items)) {
        return { type: "todo", items: raw.items };
      }
      break;
    case "shell":
      if (typeof raw.language === "string" && typeof raw.command === "string") {
        return { type: "shell", language: raw.language, command: raw.command };
      }
      break;
  }
  const { type, ...rest } = raw;
  return { type, data: rest };
});
var ToolCallSchema = z.object({
  // Always "function"
  type: z.literal("function"),
  // Tool call ID, used to match with ToolResult
  id: z.string(),
  function: z.object({
    // Tool name, e.g. "Shell", "ReadFile", "WriteFile"
    name: z.string(),
    // JSON-encoded arguments string, may be incomplete during streaming
    arguments: z.string().nullable().optional()
  }),
  // Extra metadata
  extras: z.record(z.string(), z.unknown()).nullable().optional()
});
var ToolCallPartSchema = z.object({
  // Argument chunk, appended to the last ToolCall's arguments
  arguments_part: z.string().nullable().optional()
});
var ToolReturnValueSchema = z.object({
  is_error: z.boolean(),
  output: z.union([z.string(), z.array(ContentPartSchema)]),
  message: z.string(),
  display: z.array(DisplayBlockSchema),
  extras: z.record(z.string(), z.unknown()).nullable().optional()
});
var ToolResultSchema = z.object({
  // Corresponding tool call ID
  tool_call_id: z.string(),
  return_value: ToolReturnValueSchema
});
var ClientInfoSchema = z.object({
  name: z.string(),
  version: z.string().optional()
});
var ServerInfoSchema = z.object({
  name: z.string(),
  version: z.string()
});
var SlashCommandInfoSchema = z.object({
  name: z.string(),
  description: z.string(),
  aliases: z.array(z.string())
});
var ExternalToolDefinitionSchema = z.object({
  name: z.string(),
  description: z.string(),
  parameters: z.record(z.string(), z.unknown())
});
var ExternalToolsResultSchema = z.object({
  accepted: z.array(z.string()),
  rejected: z.array(z.object({ name: z.string(), reason: z.string() }))
});
var ClientCapabilitiesSchema = z.object({
  // Whether the client supports handling QuestionRequest messages
  supports_question: z.boolean().optional()
});
var ServerCapabilitiesSchema = z.object({
  // Whether the server supports sending QuestionRequest messages
  supports_question: z.boolean().optional()
});
var InitializeParamsSchema = z.object({
  protocol_version: z.string(),
  client: ClientInfoSchema.optional(),
  external_tools: z.array(ExternalToolDefinitionSchema).optional(),
  capabilities: ClientCapabilitiesSchema.optional()
});
var InitializeResultSchema = z.object({
  protocol_version: z.string(),
  server: ServerInfoSchema,
  slash_commands: z.array(SlashCommandInfoSchema),
  external_tools: ExternalToolsResultSchema.optional(),
  capabilities: ServerCapabilitiesSchema.optional()
});
var ToolCallRequestSchema = z.object({
  id: z.string(),
  name: z.string(),
  arguments: z.string().nullable().optional()
});
var QuestionOptionSchema = z.object({
  // Option label
  label: z.string(),
  // Option description
  description: z.string().optional()
});
var QuestionItemSchema = z.object({
  // Question text
  question: z.string(),
  // Short header label (max 12 chars)
  header: z.string().optional(),
  // Options (2-4)
  options: z.array(QuestionOptionSchema),
  // Whether multiple selection is allowed
  multi_select: z.boolean().optional()
});
var QuestionRequestSchema = z.object({
  // Request ID, used for response reference
  id: z.string(),
  // Associated tool call ID
  tool_call_id: z.string(),
  // Questions (1-4)
  questions: z.array(QuestionItemSchema)
});
var QuestionResponseSchema = z.object({
  // Corresponding request ID
  request_id: z.string(),
  // Answer mapping: question text -> selected option label(s)
  answers: z.record(z.string(), z.string())
});
var TurnBeginSchema = z.object({
  // User input, can be plain text or array of content parts
  user_input: z.union([z.string(), z.array(ContentPartSchema)])
});
var StepBeginSchema = z.object({
  // Step number, starting from 1
  n: z.number()
});
var EmptyPayloadSchema = z.object({});
var StatusUpdateSchema = z.object({
  // Context usage ratio, float between 0 and 1
  context_usage: z.number().nullable().optional(),
  // Token usage stats for the current step
  token_usage: TokenUsageSchema.nullable().optional(),
  // Message ID for the current step
  message_id: z.string().nullable().optional()
});
var ApprovalRequestPayloadSchema = z.object({
  // Request ID, referenced when responding
  id: z.string(),
  // Associated tool call ID
  tool_call_id: z.string(),
  // Sender (tool name), e.g. "Shell", "WriteFile"
  sender: z.string(),
  // Action description, e.g. "run shell command"
  action: z.string(),
  // Detailed description, e.g. "Run command `rm -rf /`"
  description: z.string(),
  // Display blocks shown to the user
  display: z.array(DisplayBlockSchema).optional()
});
var ApprovalResponseEventSchema = z.object({
  // Resolved approval request ID
  request_id: z.string(),
  // Approval result
  response: ApprovalResponseSchema
});
var EventSchemas = {
  TurnBegin: TurnBeginSchema,
  TurnEnd: EmptyPayloadSchema,
  StepBegin: StepBeginSchema,
  StepInterrupted: EmptyPayloadSchema,
  CompactionBegin: EmptyPayloadSchema,
  CompactionEnd: EmptyPayloadSchema,
  StatusUpdate: StatusUpdateSchema,
  ContentPart: ContentPartSchema,
  ToolCall: ToolCallSchema,
  ToolCallPart: ToolCallPartSchema,
  ToolResult: ToolResultSchema,
  ApprovalResponse: ApprovalResponseEventSchema
};
var RequestSchemas = {
  ApprovalRequest: ApprovalRequestPayloadSchema,
  ToolCallRequest: ToolCallRequestSchema,
  QuestionRequest: QuestionRequestSchema
};
function parseEventPayload(type, payload) {
  const schema = EventSchemas[type];
  if (!schema) {
    return { ok: false, error: `Unknown event type: ${type}` };
  }
  const result = schema.safeParse(payload);
  if (!result.success) {
    return { ok: false, error: `Invalid payload for ${type}: ${result.error.message}` };
  }
  return { ok: true, value: { type, payload: result.data } };
}
function parseWireEvent(raw) {
  const result = parseEventPayload(raw.type, raw.payload);
  if (result.ok) {
    return result.value;
  }
  return {
    type: "ParseError",
    payload: {
      code: "SUBAGENT_PARSE_FAILED",
      message: result.error,
      rawType: raw.type
    }
  };
}
var SubagentEventSchema = z.lazy(
  () => z.object({
    task_tool_call_id: z.string(),
    event: z.object({ type: z.string(), payload: z.unknown() }).transform(parseWireEvent)
  })
);
EventSchemas.SubagentEvent = SubagentEventSchema;
var RunResultSchema = z.object({
  /**
   * Completion status
   * - `finished`: Turn completed normally
   * - `cancelled`: Turn was cancelled via cancel()
   * - `max_steps_reached`: Reached maximum step limit
   */
  status: z.enum(["finished", "cancelled", "max_steps_reached"]),
  // When status is max_steps_reached, returns the number of steps executed
  steps: z.number().optional()
});
var ReplayResultSchema = z.object({
  /**
   * Replay completion status
   * - `finished`: Replay completed normally
   * - `cancelled`: Replay was cancelled via cancel()
   */
  status: z.enum(["finished", "cancelled"]),
  // Number of events replayed
  events: z.number(),
  // Number of requests replayed
  requests: z.number()
});
var RpcErrorSchema = z.object({
  // Error code
  code: z.number(),
  // Error message
  message: z.string(),
  // Extra data
  data: z.unknown().optional()
});
var RpcMessageSchema = z.object({
  jsonrpc: z.string().optional(),
  id: z.string().optional(),
  method: z.string().optional(),
  params: z.unknown().optional(),
  result: z.unknown().optional(),
  error: RpcErrorSchema.optional()
});
function parseRequestPayload(type, payload) {
  const schema = RequestSchemas[type];
  if (!schema) {
    return { ok: false, error: `Unknown request type: ${type}` };
  }
  const result = schema.safeParse(payload);
  if (!result.success) {
    return { ok: false, error: `Invalid payload for ${type}: ${result.error.message}` };
  }
  return { ok: true, value: { type, payload: result.data } };
}
var ContextRecordSchema = z.object({
  role: z.string().optional(),
  content: z.unknown().optional(),
  tool_calls: z.array(
    z.object({
      id: z.string().optional(),
      function: z.object({
        name: z.string().optional(),
        arguments: z.union([z.string(), z.record(z.string(), z.unknown())]).optional()
      }).optional()
    })
  ).optional(),
  tool_call_id: z.string().optional()
});

export { ApprovalRequestPayloadSchema, ApprovalResponseEventSchema, ApprovalResponseSchema, BriefBlockSchema, ClientCapabilitiesSchema, ClientInfoSchema, ContentPartSchema, ContextRecordSchema, DiffBlockSchema, DisplayBlockSchema, EmptyPayloadSchema, EventSchemas, ExternalToolDefinitionSchema, ExternalToolsResultSchema, InitializeParamsSchema, InitializeResultSchema, QuestionItemSchema, QuestionOptionSchema, QuestionRequestSchema, QuestionResponseSchema, ReplayResultSchema, RequestSchemas, RpcErrorSchema, RpcMessageSchema, RunResultSchema, ServerCapabilitiesSchema, ServerInfoSchema, ShellBlockSchema, SlashCommandInfoSchema, StatusUpdateSchema, StepBeginSchema, SubagentEventSchema, TodoBlockSchema, TokenUsageSchema, ToolCallPartSchema, ToolCallRequestSchema, ToolCallSchema, ToolResultSchema, ToolReturnValueSchema, TurnBeginSchema, parseEventPayload, parseRequestPayload };
